File: src/main/java/pe/nanamochi/banchus/entities/db/Beatmapset.java
package pe.nanamochi.banchus.entities.db;

import jakarta.persistence.*;
import java.time.Instant;
import java.util.List;
import lombok.Data;
import lombok.ToString;
import pe.nanamochi.banchus.entities.commons.BeatmapRankedStatus;

@Entity
@Data
@ToString(exclude = "beatmaps")
@Table(
    name = "beatmapsets",
    indexes = {@Index(name = "beatmapsets_id_idx", columnList = "id")})
public class Beatmapset {
  @Id
  @Column(name = "id", nullable = false)
  private int id;

  @Column(name = "title", length = 128)
  private String title;

  @Column(name = "title_unicode", length = 128)
  private String titleUnicode;

  @Column(name = "artist", length = 128)
  private String artist;

  @Column(name = "artist_unicode", length = 128)
  private String artistUnicode;

  @Column(name = "source", length = 128)
  private String source;

  @Column(name = "source_unicode", length = 128)
  private String sourceUnicode;

  @Column(name = "creator", length = 128)
  private String creator;

  @Column(name = "tags", length = 1024)
  private String tags;

  @Column(name = "submission_status", nullable = false)
  private BeatmapRankedStatus submissionStatus;

  @Column(name = "has_video", nullable = false)
  private boolean hasVideo;

  @Column(name = "has_storyboard", nullable = false)
  private boolean hasStoryboard;

  @Column(name = "submission_date", nullable = false)
  private Instant submissionDate;

  @Column(name = "approved_date")
  private Instant approvedDate;

  @Column(name = "last_updated", nullable = false)
  private Instant lastUpdated;

  @Column(name = "total_playcount", nullable = false)
  private Long totalPlaycount;

  @Column(name = "language_id", nullable = false)
  private int languageId; // TODO: enum?

  @Column(name = "genre_id", nullable = false)
  private int genreId; // TODO: enum?

  @OneToMany(mappedBy = "beatmapset", cascade = CascadeType.ALL)
  private List<Beatmap> beatmaps;
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/beatmap/entity/Beatmapset.kt
package pe.nanamochi.banchus.beatmap.entity

import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import java.time.Instant
import pe.nanamochi.banchus.beatmap.enums.BeatmapRankedStatus

@Entity
@Table(name = "beatmapsets", indexes = [Index(name = "beatmapsets_id_idx", columnList = "id")])
class Beatmapset(
    @Id @Column(name = "id", nullable = false) var id: Int = 0,
    @Column(name = "title", length = 128) var title: String? = null,
    @Column(name = "title_unicode", length = 128) var titleUnicode: String? = null,
    @Column(name = "artist", length = 128) var artist: String? = null,
    @Column(name = "artist_unicode", length = 128) var artistUnicode: String? = null,
    @Column(name = "source", length = 128) var source: String? = null,
    @Column(name = "source_unicode", length = 128) var sourceUnicode: String? = null,
    @Column(name = "creator", length = 128) var creator: String? = null,
    @Column(name = "tags", length = 1024) var tags: String? = null,
    @Column(name = "submission_status", length = 32, nullable = false)
    @Enumerated(EnumType.STRING)
    var submissionStatus: BeatmapRankedStatus = BeatmapRankedStatus.PENDING,
    @Column(name = "has_video", nullable = false) var hasVideo: Boolean = false,
    @Column(name = "has_storyboard", nullable = false) var hasStoryboard: Boolean = false,
    @Column(name = "submission_date", nullable = false) var submissionDate: Instant = Instant.now(),
    @Column(name = "approved_date") var approvedDate: Instant? = null,
    @Column(name = "last_updated", nullable = false) var lastUpdated: Instant = Instant.now(),
    @Column(name = "total_playcount", nullable = false) var totalPlaycount: Long = 0L,
    @Column(name = "language_id", nullable = false) var languageId: Int = 0, // TODO: enum?
    @Column(name = "genre_id", nullable = false) var genreId: Int = 0, // TODO: enum?
    @OneToMany(mappedBy = "beatmapset", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    var beatmaps: MutableList<Beatmap> = mutableListOf(),
)
